### PR TITLE
add `calculate_short_bonds_given_deposit`

### PR DESCRIPTION
**Pull request by dpaiton**
# Resolved Issues
working towards https://github.com/delvtech/hyperdrive-rs/issues/185
Also makes more progress towards https://github.com/delvtech/hyperdrive-rs/issues/21

# Description
This provides an alternative function to `calculate_max_short` that includes a tolerance parameter and guarantees* convergence with sufficient number of iterations.

I changed the signature & behavior from `calculate_max_short` in these ways:
- Most importantly, we now ask for the absolute max  short (aka pool's max regardless of budget) as an input argument. This is to make clear to the user that there are two required iterative processes to get this answer: first one is to find the absolute max short, second is to find a max short relative to a budget.
- This function should also not be used with an unlimited budget to find the max short. Use `calculate_absolute_max_short` for that. If you provide a budget that is larger than the max possible, then this function will throw an error.
- The result is guaranteed to be less than or equal to your budget; it will not overshoot the target base amount even if the result is within tolerance and solvent.

 *I tested a lot of options, but I don't have any formal guarantees on convergence. The most extreme values I tried was `1e5` tolerance with `1M` steps. These two knobs can be tuned as needed for trading off compute speed against tolerance.

# Remaining tasks for future PRs
- The new test is pretty janky because of the need to put several checks around calculating the absolute max bond amount. That function needs to be investigated next.

- I still need to make the wasm & python bindings. This should be done after I fix up absolute max bond amount.

- I created this as a new function because I want to give it a chance to exist on its own and be tested before we deprecate the existing `calculate_max_short` function. Once we have battle tested it, we will want to do a cleanup.

- The method for calculating the conservative price is still not guaranteed to be safe.

### CODE REVIEW

File: crates/hyperdrive-math/src/short/max.rs [hunk 1]
```diff
@@ -38,6 +38,135 @@ impl State {
...
+
+        // Run Newton's Method to adjust the bond amount.
+        let mut loss = FixedPoint::from(U256::MAX);
+        for _ in 0..max_iterations {
+            // Calculate the current deposit.
+            let current_base_amount =
+                self.calculate_open_short(last_good_bond_amount, open_vault_share_price)?;
+
+            // Calculate the current loss.
+            // TODO: Loss should **always** go down, but it does not.
```
Comment (jrhea) [inline crates/hyperdrive-math/src/short/max.rs:91]: Should this be an issue?

File: crates/hyperdrive-math/src/short/max.rs [hunk 2]
```diff
@@ -568,9 +697,102 @@ mod tests {
...
+            let state = rng.gen::<State>();
+            let checkpoint_exposure = {
+                let value = rng.gen_range(fixed!(0)..=FixedPoint::from(U256::from(U128::MAX)));
+                if rng.gen() {
+                    -I256::try_from(value)?
+                } else {
+                    I256::try_from(value)?
+                }
+            };
+            // TODO: Absolute max doesn't always work. Sometimes the random
```
Comment (jrhea) [inline crates/hyperdrive-math/src/short/max.rs:719]: Does this need to go into an issue, or are you planning to fix this in a follow-up PR?

**Review by jrhea — APPROVED** [review]
Great work!

File: crates/hyperdrive-math/src/short/max.rs (hunk 2 above)
Comment (dpaiton) [reply crates/hyperdrive-math/src/short/max.rs:719]: good call https://github.com/delvtech/hyperdrive-rs/issues/200

File: crates/hyperdrive-math/src/short/max.rs (hunk 1 above)
Comment (dpaiton) [reply crates/hyperdrive-math/src/short/max.rs:91]: Actually I can delete this TODO. The two adjustments we make in the `Err` conditions mean that we are not truly doing Newton's Method, and thus the loss is no longer guaranteed to decrease every step.
